### PR TITLE
[RFC] Heterogeneous rank-to-GPU mapping + Qwen3.5/3.6 GGUF enablement

### PR DESCRIPTION
**Pull request by efschu**
## Summary
  Draft PR from [`efschu/shvllm@qwen35-gguf-rankgpu`](https://github.com/efschu/shvllm/tree/qwen35-gguf-rankgpu) —
  the branch behind the prebuilt image `ghcr.io/efschu/shvllm-qwen35-gguf:cu129`
  (sm75–sm120, NCCL 2.30). It bundles two features that together let a 27B
  Qwen3.6 (GGUF or HF) serve with full 262k context on 3 mismatched consumer
  GPUs (1× RTX 5090 32 GB + 2× RTX 3080 20 GB). Opening as a draft to discuss
  scope and how to split this into reviewable pieces.

  ## 1. Heterogeneous rank-to-GPU mapping

  ```bash
  nvidia-cuda-mps-control -d   # required for co-located ranks
  
  vllm serve MODEL \
    --tensor-parallel-size 4 \
    --rank-gpu-id 0,0,1,2 \          # duplicates = multiple ranks share one GPU
    --rank-gpu-memory-mib 14000      # absolute per-rank budget in MiB
  ```
    
  - Explicit TP-rank placement incl. **multiple ranks per physical GPU**
    (CUDA MPS + NCCL ≥ 2.30 `NCCL_MULTI_RANK_GPU_ENABLE`, auto-configured).
  - `--rank-gpu-memory-mib` is a single absolute MiB value, converted to a
    per-GPU fraction (`mib / nvml_total`) — the natural fit for mixed VRAM;
    mutually exclusive with `--gpu-memory-utilization`.
  - Process-level isolation (`CUDA_VISIBLE_DEVICES` per worker, NVML-resolved
    physical ids), fail-fast validation, KV-profiling serialized per GPU.
  - Scope: pure TP, single node (PP/DP/EP combinations abort with a clear
    error).

  **Correctness fixes found while validating** (independently valuable):
  - **Sampling divergence on mixed-architecture TP**: vLLM samples
    redundantly per rank and relies on bitwise-identical logits; different SM
    architectures diverge on near-tie decisions (~1/1000 tokens), silently
    desyncing rank state until outputs collapse into repetition loops. Fixed
    by broadcasting sampled + MTP-draft token ids from rank 0 when
    heterogeneous TP is detected (~KiB/step overhead). 
  - **AOT compile cache has no GPU identity in its key**: ranks on different
    architectures load each other's artifacts (hard launch failure on the
    older arch, mistuned PTX-JIT kernels on the newer). Also relevant for
    homogeneous clusters sharing a cache over NFS. Fixed by adding device
    name + capability to the hash factors.
    
  ## 2. Qwen3.5/3.6 GGUF enablement (core-side fixes)

  Small fixes so Qwen3.5/3.6 GGUF files work through the out-of-tree
  [vllm-gguf-plugin](https://github.com/vllm-project/vllm-gguf-plugin)
  (adapter lives in [a plugin fork](https://github.com/efschu/vllm-gguf-plugin/tree/qwen35-support),
  to be PR'd there separately):

  - register `Qwen3_5ForCausalLM` / `Qwen3_5MoeForCausalLM` (classes existed,
    were never registered — needed for text-only GGUF/HF checkpoints)
  - `Qwen3_5ForCausalLMBase`: add `IsHybrid` (+ mamba state classmethods) and
    `SupportsMRoPE` (text-only positions)
  - speculative config: MTP conversion for `qwen3_5_text`/`qwen3_5_moe_text`;
    draft ModelConfig inherits `model_weights` from the target
  - register the multimodal processor on `Qwen3_5MTP` (declared
    `SupportsMultiModal` but crashed MTP+vision during dummy profiling)
    
  With these + the plugin: all GGUF quant types, MTP speculative decoding from
  the GGUF's own MTP layer (62–80 % acceptance), vision via `mmproj-*.gguf`.
 
  ## Validation

  - [BENCHMARKS.md](https://github.com/efschu/shvllm/blob/qwen35-gguf-rankgpu/BENCHMARKS.md):
    6 quant variants (GGUF Q4/Q6/Q8, AWQ INT4, 2× FP8), single-stream +
    8×-parallel, up to 100k-token prefill, on the mixed 5090+3080 topology.
  - Quality: 16 long-generation batteries without repetition loops after the
    hetero-TP fix; NCCL primitive correctness verified synthetically
    (torch PG + pynccl, 0 errors in ~7k collectives on the shared-GPU
    topology).
  - Default path regression: reference launch command without the new flags
    behaves unchanged.
   
   Suggestion?

  Draft — feedback welcome before I split and polish.

### CODE REVIEW

**Comment by efschu** [comment]
Closing this one. It bundled far too much for a single review, and a history operation on my side left the branch carrying thousands of unrelated commits, so the diff here no longer reflects the actual change.

The work has been split into four separate, rebased PRs against current `main`:

- #50733 — `--rank-gpu-id` / `--rank-gpu-memory-mib`: explicit TP rank-to-GPU placement, including several ranks per physical GPU.
- #50734 — Qwen3.5 MTP fixes for text-only checkpoints.
- #50735 — `--rank-tp-ratio`: uneven tensor-parallel shards for mismatched GPUs (stacked on #50733).
- #50736 — uneven decode context parallelism (stacked on #50735).

The Qwen3.5 text-only registration part of the original description is obsolete: #50210 landed it upstream in the meantime, so #50734 only covers what is still missing.
